webrtc/webrtc.md

```diff
@@ -0,0 +1,121 @@
+   _A_ creates an SDP offer via `RTCPeerConnection.createOffer()`.
+   _A_ initiates the signaling protocol to _B_ via the relayed connection from (1), see [Signaling Protocol](#signaling-protocol) and sends the offer to _B_.
+
+5. _B_ (inbound side of relayed connection) creates an `RTCPeerConnection`.
+   Again see [STUN](#stun) section on what STUN servers to configure at creation time.
+   _B_ receives _A_'s offer sent in (2) via the signaling protocol stream and provides the offer to its `RTCPeerConnection` via `RTCPeerConnection.setRemoteDescription`.
+   _B_ then creates an answer via `RTCPeerConnection.createAnswer` and sends it to _A_ via the existing signaling protocol stream (see [Signaling Protocol](#signaling-protocol)).
+
+6. _A_ receives _B_'s answer via the signaling protocol stream and sets it locally via `RTCPeerConnection.setRemoteDescription`.
+
+7. _A_ and _B_ send their local ICE candidates via the existing signaling protocol stream.
+   Both nodes continuously read from the stream, adding incoming remote candidates via `RTCPeerConnection.addIceCandidate()`.
+
+8. On successful establishment of the direct connection, _B_ and _A_ close the signaling protocol stream.
+   On failure _B_ and _A_ reset the signaling protocol stream.
+
+   Behavior for transferring data on a relayed connection, in the case where the direct connection failed, is out of scope for this specification and dependent on the application.
+
+9. Messages on `RTCDataChannel`s on the established `RTCPeerConnection` are framed using the message framing mechanism described in [multiplexing].
+
+## STUN
+
+A node needs to discover its public IP and port, which is forwarded to the remote node in order to connect to the local node.
+On non-browser libp2p nodes doing a hole punch with TCP or QUIC, the libp2p node discovers its public address via the [identify] protocol.
+One can not use the [identify] protocol on browser nodes to discover ones public IP and port given that the browser uses a new port for each connection.
+For example say that the local browser node establishes a WebRTC connection C1 via browser-to-server to a server node and runs the [identify] protocol.
+The returned observed public port P1 will most likely (depending on the NAT) be a different port than the port observed on another connection C2.
+The only browser supported mechanism to discover ones public IP and port for a given connection is the non-libp2p protocol STUN.
+This is why this specification depends on STUN, and thus the availability of one or more STUN servers for _A_ and _B_ to discovery their public addresses.
+
+There are various publicly available STUN servers.
+As an alternative one may operate dedicated STUN servers for a given libp2p network.
+Further specification of the usage of STUN is out of scope for this specifitcation.
+
+As an aside, note that _A_ and _B_ do not need to use the same STUN server in order to establish a direct WebRTC connection.
```

Comment:
Nit:
```suggestion
It is not necessary that _A_ and _B_ do not need to use the same STUN server in order to establish a direct WebRTC connection.
```

Reply:
Adopted the first as it stresses that there is no reason to use the same one, the latter sounds to me like it is an option to use different ones.

3425f01

Reply:
Can we at least reduce the negation please:

```suggestion
It is not necessary for _A_ and _B_ to use the same STUN server when establishing a WebRTC connection.
```

Reply:
:+1: done in 554e3db. Thanks.